File: files/CH-192.R
library(tidyverse)
library(readxl)

path = "files/CH-192 Table Transformation.xlsx"
input = read_excel(path, range = "B2:D12")
test  = read_excel(path, range = "B15:D18")

result = input %>%
  mutate(sign = sign(Quantity)) %>%
  mutate(group = ceiling(cumsum(sign != lag(sign, default = 0))/2), .by = Product) %>%
  summarise(Quantity = sum(Quantity), 
            Date = min(Date),
            .by = c(Product,group)) %>%
  filter(Quantity != 0) %>%
  select(Date, Product, Quantity)

all.equal(result, test)
#> [1] TRUE

File: files/CH-192.py
import pandas as pd
import numpy as np

path = "CH-192 Table Transformation.xlsx"
input = pd.read_excel(path, usecols="B:D", skiprows=1, nrows=11)
test = pd.read_excel(path, usecols="B:D", skiprows=14, nrows=4)

input['sign'] = np.sign(input['Quantity'])
input['group'] = input.groupby('Product')['sign'].apply(lambda x: (x != x.shift()).cumsum()).reset_index(level=0, drop=True)
input['group'] = np.ceil(input['group']/2)
result = input.groupby(['Product', 'group']).agg({'Quantity': 'sum', 'Date': 'min'}).reset_index()
result = result[result['Quantity'] != 0][['Date', 'Product', 'Quantity']].sort_values(by = "Date").reset_index(drop= True)

print(result.equals(test)) # True